core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -138,12 +138,13 @@ private void internalAddColumn(String parent, String name, boolean isOptional, T
       parentId = parentField.fieldId();
       Preconditions.checkArgument(!deletes.contains(parentId),
           "Cannot add to a column that will be deleted: %s", parent);
-      Preconditions.checkArgument(schema.findField(parent + "." + name) == null,
-          "Cannot add column, name already exists: %s.%s", parent, name);
+      Preconditions.checkArgument(schema.findField(parent + "." + name) == null ||
+                      deletes.contains(schema.findField(parent + "." + name).fieldId()),
+          "Cannot add column, name already exists and is not being deleted: %s.%s", parent, name);
```

Comment:
I don't think the error message needs to change. It isn't helpful in most cases to add the "and is not being deleted". That makes it harder to understand when adding a deleted column is a rare case.